services/agent/src/simcore_service_agent/services/containers_manager.py

```diff
@@ -0,0 +1,83 @@
+import logging
+from dataclasses import dataclass, field
+
+from aiodocker import Docker
+from fastapi import FastAPI
+from models_library.api_schemas_directorv2.services import (
+    DYNAMIC_PROXY_SERVICE_PREFIX,
+    DYNAMIC_SIDECAR_SERVICE_PREFIX,
+)
+from models_library.projects_nodes_io import NodeID
+from servicelib.fastapi.app_state import SingletonInAppStateMixin
+from servicelib.utils import limited_gather
+
+from ..core.settings import ApplicationSettings
+from .docker_utils import get_containers_with_prefixes, remove_container_forcefully
+
+_logger = logging.getLogger(__name__)
+_always_visible_logger = logging.getLogger(__name__)
+
+
+@dataclass
+class ContainersManager(SingletonInAppStateMixin):
+    app_state_name: str = "containers_manager"
+
+    docker: Docker = field(default_factory=Docker)
+
+    async def force_container_cleanup(self, node_id: NodeID) -> None:
+        # compose all possible used container prefixes
+        proxy_prefix = f"{DYNAMIC_PROXY_SERVICE_PREFIX}_{node_id}"
+        dy_sidecar_prefix = f"{DYNAMIC_SIDECAR_SERVICE_PREFIX}_{node_id}"
+        user_service_prefix = f"{DYNAMIC_SIDECAR_SERVICE_PREFIX}-{node_id}"
+
+        orphan_containers = await get_containers_with_prefixes(
+            self.docker, {proxy_prefix, dy_sidecar_prefix, user_service_prefix}
+        )
+        _logger.debug(
+            "Detected orphan containers for node_id='%s': %s",
+            node_id,
+            orphan_containers,
+        )
+
+        unexpected_orphans = {
+            orphan
+            for orphan in orphan_containers
+            if orphan.startswith(user_service_prefix)
+        }
+        if unexpected_orphans:
+            _always_visible_logger.info(
+                "Unexpected orphans detected for node_id='%s': %s",
+                node_id,
+                unexpected_orphans,
+            )
+
+        await limited_gather(
```

Comment:
for info limited_gather by default runs 1 task at a time... so not sure it is very useful here ;)
you can use the generator version otherwise with ```async for```

Reply:
I think it's better to go one request at a time in this situation. The parallel version is not useful. Will replace.